ArduPlane/system.cpp

```diff
@@ -318,200 +289,39 @@ void Plane::set_mode(enum FlightMode mode, mode_reason_t reason)
-        // for ArduSoar soaring_controller
-        g2.soaring_controller.init_cruising();
-#endif
-        
-        set_target_altitude_current();
-        break;
-
-    case FLY_BY_WIRE_B:
-        throttle_allows_nudging = false;
-        auto_throttle_mode = true;
-        auto_navigation_mode = false;
-
-#if SOARING_ENABLED == ENABLED
-        // for ArduSoar soaring_controller
-        g2.soaring_controller.init_cruising();
-#endif
-
-        set_target_altitude_current();
-        break;
-
-    case CIRCLE:
-        // the altitude to circle at is taken from the current altitude
-        throttle_allows_nudging = false;
-        auto_throttle_mode = true;
-        auto_navigation_mode = true;
-        next_WP_loc.alt = current_loc.alt;
-        break;
-
-    case AUTO:
-        throttle_allows_nudging = true;
-        auto_throttle_mode = true;
-        auto_navigation_mode = true;
-        if (quadplane.available() && quadplane.enable == 2) {
-            auto_state.vtol_mode = true;
-        } else {
-            auto_state.vtol_mode = false;
-        }
-        next_WP_loc = prev_WP_loc = current_loc;
-        // start or resume the mission, based on MIS_AUTORESET
-        mission.start_or_resume();
-
-#if SOARING_ENABLED == ENABLED
-        g2.soaring_controller.init_cruising();
-#endif
-        break;
-
-    case RTL:
-        throttle_allows_nudging = true;
-        auto_throttle_mode = true;
-        auto_navigation_mode = true;
-        prev_WP_loc = current_loc;
-        do_RTL(get_RTL_altitude());
-        break;
-
-    case LOITER:
-        throttle_allows_nudging = true;
-        auto_throttle_mode = true;
-        auto_navigation_mode = true;
-        do_loiter_at_location();
-
-#if SOARING_ENABLED == ENABLED		
-        if (g2.soaring_controller.is_active() &&
-            g2.soaring_controller.suppress_throttle()) {
-			g2.soaring_controller.init_thermalling();
-			g2.soaring_controller.get_target(next_WP_loc); // ahead on flight path
-		}
-#endif
-		
-        break;
-
-    case AVOID_ADSB:
-    case GUIDED:
-        throttle_allows_nudging = true;
-        auto_throttle_mode = true;
-        auto_navigation_mode = true;
-        guided_throttle_passthru = false;
-        /*
-          when entering guided mode we set the target as the current
-          location. This matches the behaviour of the copter code
-        */
-        guided_WP_loc = current_loc;
-        set_guided_WP();
-        break;
-
-    case QSTABILIZE:
-    case QHOVER:
-    case QLOITER:
-    case QLAND:
-    case QRTL:
-    case QAUTOTUNE:
-    case QACRO:
-        throttle_allows_nudging = true;
-        auto_navigation_mode = false;
-        if (!quadplane.init_mode()) {
-            control_mode = previous_mode;
-        } else {
-            auto_throttle_mode = false;
-            auto_state.vtol_mode = true;
-        }
-        break;
+        // we failed entering new mode, roll back to old
+        previous_mode = &old_previous_mode;
```

Comment:
Don't you need to re-enter the old mode, as enter a mode has significant side effects? (IE navigation target etc).

Reply:
yeah, quadplane is the black duck here. It's the only thing that can fail a set_mode. I don't have a quadplane to test thoroughly. Can you test it for me?

Reply:
@magicrub, I'd vote for testing it in SITL..